### PR TITLE
schemaworkload: Add metrics that keep track of workload operations

### LINKED ISSUE #81928: workload/schemachange: sanity check that the workload did things
**Is your feature request related to a problem? Please describe.**

Generally I have a fear that you can break the workload in such a way that it looks like it keeps passing but it stops actually doing interesting things. This happened long ago where it put the transaction into a bad state and then ignored the errors. 

**Describe the solution you'd like**
We should sanity check some combination of counters, session metrics, or perhaps quantifying the size of the produced schema graph in order to ensure that stuff did happen.

There's lots of different ways to approach this.

Jira issue: CRDB-16339

Epic CRDB-19168

### PR DESCRIPTION
**Pull request by rimadeodhar**
This PR adds metrics that indicate operations performed by the schema change workload. These metrics are output in logs at the end of a workload run. These metrics will function as a sanity check to ensure that the workload is doing something meaningful and will need to be checked manually. The current set of metrics are fairly rudimentary but will help provide some initial insight into the efficacy of this workload.

Fixes: https://github.com/cockroachdb/cockroach/issues/81928
Epic: none
Release note: None

### CODE REVIEW

**Review by fqazi — APPROVED** [review]
Just a few minor comments, thanks for doing this.


:lgtm:

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117743#-:-NowpliJ22-C67vRyc8k:b-75vhz2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @rimadeodhar)

___
*[`pkg/workload/schemachange/schemachange.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117743#-NowpUOVBMt9zSIwlL7z:-NowpUOVBMt9zSIwlL8-:b-z2wgbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e76e14711f0608c3d9f708c4d5a8e22fc90c9dba/pkg/workload/schemachange/schemachange.go#L627)):*
> ```Go
> 	w.logger.flushLog("")
> 	w.recordInHist(timeutil.Since(start), txnOk)
> 	workloadMetrics[txnCommitted] = attribute.BoolValue(true)
> ```

This feels a bit leaky should we just have a function to set the commit state?
___
*[`pkg/workload/schemachange/workload_result.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117743#-Nowo6GkECG8_2bYF8ze:-Nowo6GkECG8_2bYF8zf:b-wxpij2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e76e14711f0608c3d9f708c4d5a8e22fc90c9dba/pkg/workload/schemachange/workload_result.go#L73)):*
> ```Go
> 	if sp.Name() == schemaWorkerSpanName {
> 		for _, attr := range sp.Attributes() {
> 			fmt.Printf("Attribute key %s\n", attr.Key)
> ```

Left over debug code

File: pkg/workload/schemachange/schemachange.go [hunk 1]
```diff
@@ -585,26 +604,27 @@ func (w *schemaChangeWorker) run(ctx context.Context) error {
...
+		w.logger.flushLogWithError(err)
 		w.preErrorHook()
 		return errors.Mark(err, errRunInTxnFatalSentinel)
 	}
 
 	// If there were no errors while committing the txn.
-	w.logger.flushLog(tx, "")
+	w.logger.flushLog("")
 	w.recordInHist(timeutil.Since(start), txnOk)
+	workloadMetrics[txnCommitted] = attribute.BoolValue(true)
```
Comment (Xiang-Gu) [inline pkg/workload/schemachange/schemachange.go:627]: Is this `workloadMetrics` capturing all transactions ever run within one worload worker? If so, why is "txnCommitted" a boolean valued attribute that is either `True` or `False`?

I though a worker could perform many transactions, some succeeded and some failed, so we ought to store the number of succeeded/failed txns ever performed within this work.

But I felt like I missed something here....

**Review by rimadeodhar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117743)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi, @herkolategan, @renatolabs, and @Xiang-Gu)

___
*[`pkg/workload/schemachange/schemachange.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117743#-NowpUOVBMt9zSIwlL7z:-Npub-YG0lRnOzIF4pc-:b-oi6ogs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e76e14711f0608c3d9f708c4d5a8e22fc90c9dba/pkg/workload/schemachange/schemachange.go#L627)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This feels a bit leaky should we just have a function to set the commit state?
</blockquote></details>

Hmm, could you elaborate? I'm not sure as to why this would be leaky?
___
*[`pkg/workload/schemachange/schemachange.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117743#gh-1465454646:-NpuceDk7vadwp-vQ5py:b-8gh6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e76e14711f0608c3d9f708c4d5a8e22fc90c9dba/pkg/workload/schemachange/schemachange.go#L627)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Is this `workloadMetrics` capturing all transactions ever run within one worload worker? If so, why is "txnCommitted" a boolean valued attribute that is either `True` or `False`?

I though a worker could perform many transactions, some succeeded and some failed, so we ought to store the number of succeeded/failed txns ever performed within this work.

But I felt like I missed something here....
</blockquote></details>

Not quite. Every worker runs a set of schema operations within one transaction. The workload uses multiple workers to run multiple concurrent transactions. Additionally, the top level workload setup can also have a single worker run multiple txns serially as in https://github.com/cockroachdb/cockroach/blob/82e2348aac880be042fbd8533d734693a69e434b/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go#L121. 
The way the metric collection is set up is that each workload metric collection span corresponds to the single transaction (run() method) that each schema change worker runs. These attributes are aggregated when every "span" terminates (at the end of the run() method). These aggregations are stored within a span processor which is only initialized once for a workload.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117743)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @renatolabs, @rimadeodhar, and @Xiang-Gu)

___
*[`pkg/workload/schemachange/schemachange.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117743#-NowpUOVBMt9zSIwlL7z:-NpudXha5C5g2lpv56du:b-pmgoc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e76e14711f0608c3d9f708c4d5a8e22fc90c9dba/pkg/workload/schemachange/schemachange.go#L627)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Hmm, could you elaborate? I'm not sure as to why this would be leaky?
</blockquote></details>

I just meant we have wrappers for other operations on workloadMetrics versus directly accessing the map. But, I don't feel strongly about this.

**Review by rimadeodhar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117743)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi, @herkolategan, @renatolabs, and @Xiang-Gu)

___
*[`pkg/workload/schemachange/schemachange.go` line 627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117743#-NowpUOVBMt9zSIwlL7z:-Npv0_Cj6jw-CO1zIhxh:bl118g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e76e14711f0608c3d9f708c4d5a8e22fc90c9dba/pkg/workload/schemachange/schemachange.go#L627)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I just meant we have wrappers for other operations on workloadMetrics versus directly accessing the map. But, I don't feel strongly about this.
</blockquote></details>

Gotcha. Yeah, it didn't make sense to add a wrapper for one line code only used in one place. We can always refactor in the future. Leaving this in as is for now.

**Comment by rimadeodhar** [comment]
TFTR!

bors r=fqazi

**Pull request merged**